src/storage/txn/flow_controller.rs

```diff
@@ -0,0 +1,1055 @@
+const SPARE_TICKS_THRESHOLD: u64 = 10;
+const RATIO_SCALE_FACTOR: f64 = 10000000.0;
+const EMA_FACTOR: f64 = 0.6; // EMA stands for Exponential Moving Average
+const LIMIT_UP_PERCENT: f64 = 0.04; // 4%
+const LIMIT_DOWN_PERCENT: f64 = 0.02; // 2%
+const MIN_THROTTLE_SPEED: f64 = 16.0 * 1024.0; // 16KB
+const MAX_THROTTLE_SPEED: f64 = 200.0 * 1024.0 * 1024.0; // 200MB
+
+const PID_KP_FACTOR: f64 = 0.15;
+const PID_KD_FACTOR: f64 = 5.0;
+
+#[derive(Eq, PartialEq)]
+enum Trend {
+    Increasing,
+    Decreasing,
+    NoTrend,
+}
+
+// Flow controller is used to throttle the write rate at scheduler level, aiming
+// to substitute the write stall mechanism of RocksDB. It features in two points:
+//   * throttle at scheduler, so raftstore and apply won't be blocked anymore
+//   * better control on the throttle rate to avoid QPS drop under heavy write
+//
+// When write stall happens, the max speed of write rate max_delayed_write_rate
+// is limited to 16MB/s by default which doesn't take real disk ability into
+// account. It may underestimate the disk's throughout that 16MB/s is too small
+// at once, causing a very large jitter on the write duration.
+// Also, it decreases the delayed write rate further if the factors still exceed
+// the threshold. So under heavy write load, the write rate may be throttled to
+// a very low rate from time to time, causing QPS drop eventually.
+//
+// The main idea of this flow controller is to throttle at a steady write rate
+// so that the number of L0 keeps around the threshold. When it falls below the
+// threshold, the throttle state wouldn't exit right away. Instead, it may keep
+// or increase the throttle speed depending on some statistics.
+//
+// How can we decide the throttle speed?
+// It uses 95th write rate of the last few seconds as the initial throttle speed.
+// Then as we can imagine, the consumption ability of L0 wouldn't change
+// dramatically corresponding to the ability of hardware. So we can record the
+// flush flow when reaching the threshold as target flow, and increase or
+// decrease the throttle speed based on whether current flush flow is smaller or
+// larger than target flow.
+
+// For compaction pending bytes, we use discardable ratio to do flow control
+// which is separated mechanism from throttle speed. Compaction pending bytes is
+// a approximate value, usually, changes up and down dramatically, so it's unwise
+// to map compaction pending bytes to a specified throttle speed. Instead,
+// mapping it from soft limit to hard limit as 0% to 100% discardable ratio. With
+// this, there must be a point that foreground write rate is equal to the
+// background compaction pending bytes consuming rate so that compaction pending
+// bytes is kept around a steady level.
+//
+// Here is a brief flow showing where the mechanism works:
+// grpc -> check should drop(discardable ratio) -> limiter -> async write to raftstore
+//
```

Comment:
Comment is a little too long.. Maybe simplify a little? or add some more bullets.

Reply:
Personally speaking, not a bad idea for a module introduction.

Reply:
Probably, then make it doc comments?